File: PHRPReader/clsMSGFResultsReader.vb
'*********************************************************************************************************
' Written by Matthew Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
'
' Created 04/03/2012
'
' This class reads MSGF scores from a tab-delimited _msgf.txt file
'
'*********************************************************************************************************

Option Strict On
Imports System.IO

Public Class clsMSGFResultsReader

#Region "Constants"
	Public Const DATA_COLUMN_ResultID As String = "Result_ID"
	Public Const DATA_COLUMN_Scan As String = "Scan"
	Public Const DATA_COLUMN_Charge As String = "Charge"
	Public Const DATA_COLUMN_Protein As String = "Protein"
	Public Const DATA_COLUMN_Peptide As String = "Peptide"
	Public Const DATA_COLUMN_SpecProb As String = "SpecProb"
	Public Const DATA_COLUMN_Notes As String = "Notes"
#End Region

#Region "Class-wide variables"
	' Column headers
    Private ReadOnly mColumnHeaders As SortedDictionary(Of String, Integer)
    Private mErrorMessage As String = String.Empty
#End Region

	''' <summary>
	''' Error message
	''' </summary>
	Public ReadOnly Property ErrorMessage As String
		Get
			If String.IsNullOrEmpty(mErrorMessage) Then
				Return String.Empty
			Else
				Return mErrorMessage
			End If
		End Get
	End Property

	''' <summary>
	''' Constructor
	''' </summary>
	''' <remarks></remarks>
	Public Sub New()
		mColumnHeaders = New SortedDictionary(Of String, Integer)
	End Sub

    Private Sub AddHeaderColumn(strColumnName As String)
        mColumnHeaders.Add(strColumnName, mColumnHeaders.Count)
    End Sub

    Private Sub DefineColumnHeaders()

        mColumnHeaders.Clear()

        ' Define the default column mapping
        AddHeaderColumn(DATA_COLUMN_ResultID)
        AddHeaderColumn(DATA_COLUMN_Scan)
        AddHeaderColumn(DATA_COLUMN_Charge)
        AddHeaderColumn(DATA_COLUMN_Protein)
        AddHeaderColumn(DATA_COLUMN_Peptide)
        AddHeaderColumn(DATA_COLUMN_SpecProb)
        AddHeaderColumn(DATA_COLUMN_Notes)

    End Sub

    ''' <summary>
    ''' Open a tab-delimited MSGF results file and read the data
    ''' </summary>
    ''' <param name="strInputFilePath">Input file path</param>
    ''' <returns>A Dictionary where keys are ResultID and values are MSGF_SpecProb values (stored as strings)</returns>
    Public Function ReadMSGFData(strInputFilePath As String) As Dictionary(Of Integer, String)

        Dim lstMSGFData As Dictionary(Of Integer, String)
        lstMSGFData = New Dictionary(Of Integer, String)

        Dim strLineIn As String
        Dim strSplitLine() As String
        Dim blnHeaderLineParsed As Boolean
        Dim blnSkipLine As Boolean

        Dim intLinesRead As Integer
        Dim intResultID As Integer
        Dim strMSGFSpecProb As String

        Try
            DefineColumnHeaders()
            intLinesRead = 0
            mErrorMessage = String.Empty

            Using srInFile = New StreamReader(New FileStream(strInputFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
                Do While Not srInFile.EndOfStream
                    strLineIn = srInFile.ReadLine()
                    intLinesRead += 1
                    blnSkipLine = False

                    If Not String.IsNullOrWhiteSpace(strLineIn) Then
                        strSplitLine = strLineIn.Split(ControlChars.Tab)

                        If Not blnHeaderLineParsed Then
                            If Not clsPHRPReader.IsNumber(strSplitLine(0)) Then
                                ' Parse the header line to confirm the column ordering
                                clsPHRPReader.ParseColumnHeaders(strSplitLine, mColumnHeaders)
                                blnSkipLine = True
                            End If

                            blnHeaderLineParsed = True
                        End If

                        If Not blnSkipLine AndAlso strSplitLine.Length >= 4 Then

                            intResultID = clsPHRPReader.LookupColumnValue(strSplitLine, DATA_COLUMN_ResultID, mColumnHeaders, -1)

                            If intResultID >= 0 Then
                                strMSGFSpecProb = clsPHRPReader.LookupColumnValue(strSplitLine, DATA_COLUMN_SpecProb, mColumnHeaders)

                                If Not String.IsNullOrEmpty(strMSGFSpecProb) AndAlso Not lstMSGFData.ContainsKey(intResultID) Then
                                    lstMSGFData.Add(intResultID, strMSGFSpecProb)
                                End If
                            End If

                        End If
                    End If

                Loop
            End Using

        Catch ex As Exception
            mErrorMessage = "Error reading the MSGF data: " & ex.Message
        End Try

        Return lstMSGFData
    End Function

End Class

File: PHRPReader/clsMSGFResultsReader.cs
//*********************************************************************************************************
// Written by Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
//
// Created 04/03/2012
//
//*********************************************************************************************************

using System;
using System.Collections.Generic;
using System.IO;

namespace PHRPReader
{
    /// <summary>
    /// This class reads MSGF scores from a tab-delimited _msgf.txt file
    /// </summary>
    public class clsMSGFResultsReader
    {
        #region "Constants"

#pragma warning disable 1591
        public const string DATA_COLUMN_ResultID = "Result_ID";
        public const string DATA_COLUMN_Scan = "Scan";
        public const string DATA_COLUMN_Charge = "Charge";
        public const string DATA_COLUMN_Protein = "Protein";
        public const string DATA_COLUMN_Peptide = "Peptide";
        public const string DATA_COLUMN_SpecProb = "SpecProb";
        public const string DATA_COLUMN_Notes = "Notes";
#pragma warning restore 1591

        #endregion

        #region "Class-wide variables"

        /// <summary>
        /// Column headers
        /// </summary>
        private readonly SortedDictionary<string, int> mColumnHeaders;

        private string mErrorMessage = string.Empty;

        #endregion

        /// <summary>
        /// Error message
        /// </summary>
        public string ErrorMessage
        {
            get
            {
                if (string.IsNullOrEmpty(mErrorMessage))
                {
                    return string.Empty;
                }

                return mErrorMessage;
            }
        }

        /// <summary>
        /// Constructor
        /// </summary>
        public clsMSGFResultsReader()
        {
            mColumnHeaders = new SortedDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        }

        private void AddHeaderColumn(string columnName)
        {
            mColumnHeaders.Add(columnName, mColumnHeaders.Count);
        }

        /// <summary>
        /// Define header names for MSGF result files
        /// </summary>
        private void DefineColumnHeaders()
        {
            mColumnHeaders.Clear();

            // Define the default column mapping
            AddHeaderColumn(DATA_COLUMN_ResultID);
            AddHeaderColumn(DATA_COLUMN_Scan);
            AddHeaderColumn(DATA_COLUMN_Charge);
            AddHeaderColumn(DATA_COLUMN_Protein);
            AddHeaderColumn(DATA_COLUMN_Peptide);
            AddHeaderColumn(DATA_COLUMN_SpecProb);
            AddHeaderColumn(DATA_COLUMN_Notes);
        }

        /// <summary>
        /// Open a tab-delimited MSGF results file and read the data
        /// </summary>
        /// <param name="inputFilePath">Input file path</param>
        /// <returns>A Dictionary where keys are ResultID and values are MSGF_SpecProb values (stored as strings)</returns>
        public Dictionary<int, string> ReadMSGFData(string inputFilePath)
        {
            var msgfData = new Dictionary<int, string>();

            try
            {
                DefineColumnHeaders();

                mErrorMessage = string.Empty;

                using (var reader = new StreamReader(new FileStream(inputFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)))
                {
                    var headerLineParsed = false;

                    while (!reader.EndOfStream)
                    {
                        var lineIn = reader.ReadLine();
                        var skipLine = false;

                        if (string.IsNullOrWhiteSpace(lineIn))
                            continue;

                        var splitLine = lineIn.Split('\t');

                        if (!headerLineParsed)
                        {
                            if (!clsPHRPReader.IsNumber(splitLine[0]))
                            {
                                // Parse the header line to confirm the column ordering
                                clsPHRPReader.ParseColumnHeaders(splitLine, mColumnHeaders);
                                skipLine = true;
                            }

                            headerLineParsed = true;
                        }

                        if (!skipLine && splitLine.Length >= 4)
                        {
                            var resultID = clsPHRPReader.LookupColumnValue(splitLine, DATA_COLUMN_ResultID, mColumnHeaders, -1);

                            if (resultID >= 0)
                            {
                                var msgfSpecProb = clsPHRPReader.LookupColumnValue(splitLine, DATA_COLUMN_SpecProb, mColumnHeaders);

                                if (!string.IsNullOrEmpty(msgfSpecProb) && !msgfData.ContainsKey(resultID))
                                {
                                    msgfData.Add(resultID, msgfSpecProb);
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                mErrorMessage = "Error reading the MSGF data: " + ex.Message;
            }

            return msgfData;
        }
    }
}
